src/psp.ts

```diff
@@ -0,0 +1,502 @@
+          // Check conditions
+          // **************************************************************************
+          node.logger.debug('Checking conditions...', subscriptionIdLogOptions);
+
+          // Is this step really needed?
+          // https://github.com/api3dao/airnode/blob/v1-protocol/packages/airnode-protocol-v1/contracts/dapis/DapiServer.sol#L20-L23
+
+          const encodedFulfillmentData = ethers.utils.defaultAbiCoder.encode(['int256'], [apiValue]);
+
+          // TODO: Should "subscription.conditions" be already encoded in config?
+          // const encodedConditionParameters = ethers.utils.defaultAbiCoder.encode(
+          //   ['uint256'],
+          //   [subscription.conditions] // How is this supposed to be set in config? i.e. (10**18 / 10) for 10%?
+          // );
+
+          const voidSigner = new ethers.VoidSigner(ethers.constants.AddressZero, provider);
+          // TODO: retryGo?
+          if (
+            !(await dapiServer
+              .connect(voidSigner)
+              .conditionPspBeaconUpdate(subscriptionId, encodedFulfillmentData, subscription.conditions))
+          ) {
+            node.logger.warn('Conditions not met. Skipping update...', subscriptionIdLogOptions);
+            continue;
+          }
+          node.logger.info('Conditions met. Updating beacon...', subscriptionIdLogOptions);
+
+          // **************************************************************************
+          // Compute signature
+          // **************************************************************************
+          node.logger.debug('Signing fulfill message...', subscriptionIdLogOptions);
+
+          const airnodeWallet = ethers.Wallet.fromMnemonic(nodeSettings.airnodeWalletMnemonic).connect(provider);
+          // TODO: protocol parameter is missing in v0.3.1
+          // const sponsorWallet = node.evm.deriveSponsorWallet(airnodeHDNode, subscription.sponsor).connect(provider);
+          const sponsorWallet = deriveSponsorWallet(
+            nodeSettings.airnodeWalletMnemonic,
+            subscription.sponsor,
+            '2'
+          ).connect(provider);
+
+          const timestamp = nowInSeconds();
+
+          const signature = await airnodeWallet.signMessage(
+            ethers.utils.arrayify(
+              ethers.utils.keccak256(
+                ethers.utils.solidityPack(
+                  ['bytes32', 'uint256', 'address'],
+                  [subscriptionId, timestamp, sponsorWallet.address]
+                )
+              )
+            )
+          );
+
+          // **************************************************************************
+          // Update beacon
+          // **************************************************************************
+          node.logger.debug('Fulfilling subscription...', subscriptionIdLogOptions);
+
+          // TODO: what if there is a previous pending fulfillPspBeaconUpdate (or any pending?) tx for this sponsor wallet?
```

Comment:
You should derive sponsor wallet addresses at the beginning and get the transaction counts immediately. Then you make the fulfillment transactions at the end by incrementing that (like Airnode). This means repeated fulfillments will have the same nonce and cancel each other out.

Reply:
Oh right, I always forget about setting the nonce. I'm already doing it for current Airkeeper implementation: https://github.com/api3dao/airkeeper/blob/psp-beacon-update/src/start.ts#L161-L176